File: src/go/problems/generate_parenthesis_test.go
// https://leetcode.com/problems/generate-parentheses/

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type GenerateParenthesis struct{}

func TestGenerateParenthesis(t *testing.T) {
	gen := core.TestSuite[GenerateParenthesis]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(3).Result([]string{"((()))", "(()())", "(())()", "()(())", "()()()"})
	}).Add(func(tc *core.TestCase) {
		tc.Param(1).Result([]string{"()"})
	}).Run(t)
}

func (GenerateParenthesis) Solution(n int) []string {
	result := make([]string, 0)

	build(n-1, n, "(", "", &result)
	return result
}

func build(open int, closed int, parenthesis string, temp string, result *[]string) {
	temp += parenthesis

	if open > 0 {
		build(open-1, closed, "(", temp, result)
	}

	if closed > open {
		build(open, closed-1, ")", temp, result)
	}

	if open == 0 && closed == 0 {
		*result = append(*result, temp)
	}
}

File: src/python/problems/generate_parenthesis_test.py
#https://leetcode.com/problems/generate-parentheses/
from core.problem_base import *

class GenerateParenthesis(ProblemBase):
    def Solution(self, n: int) -> list[str]:
        result = []
        def Build(open, closed, parenthesis, temp):
            temp += parenthesis

            if open > 0:
                Build(open - 1, closed, '(', temp)

            if closed > open:
                Build(open, closed - 1, ')', temp)

            if open == 0 and closed == 0:
                result.append(temp)

            temp = temp[:-1]

        Build(n - 1, n, '(', "")
        return result


if __name__ == '__main__':
    TestGen(GenerateParenthesis) \
        .Add(lambda tc: tc.Param(3).Result(["((()))","(()())","(())()","()(())","()()()"])) \
        .Add(lambda tc: tc.Param(1).Result(["()"])) \
        .Run()
